rfcs/20240326-block-level-api/README.md

```diff
@@ -0,0 +1,441 @@
+reference], and is currently used in oneDNN CPU backend to implement
+Matmul, Convolution, Inner product and RNN primitives.  In particular,
+the main operation batch-reduce GEneral Matrix Multiply operation (aka
+brgemm), is a very flexible operation that can be used in a variety of
+DNN operators. It is defined as folllow:
+
+![](brgemm_pic.png)
+```math
+D = \beta C + \alpha \sum_i A_i \cdot B_i + bias
+```
+
+with 
+- $A_i$ a set of matrices of dimension $MxK$
+- $B_i$ a set of matrices of dimension $KxN$ 
+- D and C matrices of dimension $MxN$
+- bias a vector of dimension $N$.
+
+This proposal discusses exposing these sequential, basic building
+blocks externally for end-users to conveniently be able to write their
+own optimized custom primitives/routines. These can be used in their
+own applications and/or contributed back behind the oneDNN graph API.
+
+As a starter, we will expose only simple transforms and brgemm
+routines.
+
+## Memory descriptors, memory objects and/or pointers ?
+
+In general, we need to express memory shapes and memory objects, and
+for this block level API, we will have are three considerations:
+- allowing arbitrary strides between blocks (A,B in particular) for
+  maximum flexibility,
+- Support for metadata (e.g. zero-point handling) and
+  hardware-specific layouts,
+- runtime overheads.
+
+### Arbitrary strides between A and B blocks
+
+To make the API flexible, we want to allow arbitrary strides between A
+and B blocks. This is unfortunately not possible if we encapsulate all
+A blocks (resp B blocks) of a brgemm operation in a dnnl::memory
+object, since this will force a fixed stride between each block.
+
+Another option would be to take an array of dnnl::memory objects, each
+containing a single block of A/B. This would require the user to call
+`dnnl::memory::set_data_handle()` on each of these objects before
+brgemm call. It would incur some unnecessary overhead, but should
+not be anything substantial.
+
+However, one might wonder why use memory descriptors at all and not
+just take pointers directly as input to brgemm::execute
+call. Internally we support multiple flavors:
+- array of pointers: here the user will have to do pointer arithmetic
+  before passing those pointers to brgemm::execute call.
+- base pointer + arrray of offsets: here the pointer arithmetic will
+  be handled internally behind brgemm::execute call.
+- base pointer + fixed stride: this flavor has limited application and
+  is typically used for matmul implementation. Given its limited
+  applicability, it seems unnecessary to expose it.
+  
+To make the public brgemm API simpler, we propose to expose only the array
+of pointers flavor.
+
+For the brgemm API specifically, an array of equal sizes for A and B
+blocks are passed.  To guarentee these are of equal size, and simplify
+pointer arithmetic logic inside generated kernels, we will take an
+array of pairs of pointers.
```

Comment:
Correct. In general:
- we can't use single memory object as it would constrain the strides between $A_i$ and $B_i$ blocks.
- we could take an array of memory objects, but it would not add value since all blocks should have the same descriptor.

Reply:
I see. Thanks for explaining.